blog/2020-02-16-debezium-on-rds.adoc

```diff
@@ -0,0 +1,226 @@
+
+There were a few mistakes we made which we learnt about during lots of stress-testing and failover testing. All of the solutions to our mistakes are already documented in the Debezium documentation but they are all listed together here to make it difficult to repeat our mistakes.
+
+* We initially used the *wal2json* plugin which caused the connector to encounter *OutOfMemoryError* when committing large transactions. Hence our recommendation is:
+** On PostgreSQL < 10, use the *wal2json_streaming* plugin to avoid *OutOfMemoryError* on large transactions.
+** On PostgreSQL >= 10, use the *pgoutput* plugin.
+* We were producing JSON messages with schemas enabled; this creates larger Kafka records than needed, in particular if schema changes are rare. Hence we decided to disable message schemas by setting `key.converter.schemas.enabled` and `value.converter.schemas.enabled` to `false`.
+* We were observing a lot of columns with base64 encoded data. As documented, that's the default, but can be hard to handle for consumers.  To convert to a format that's easier to parse at the expense of some accuracy loss configure the data type specific properties https://debezium.io/documentation/reference/1.0/connectors/postgresql.html#data-types[as documented]. Specifically, set `decimal.handling.mode` to `string` to receive NUMERIC, DECIMAL and equivalent types as a string (eg. "3.14") and set `hstore.handling.mode` to `json` to receive HSTORE columns as a JSON string.
+* No alarms were set up on the transaction logs disk usage on the database. We added alarms on the RDS metric *TransactionLogsDiskUsage* and *OldestReplicationSlotLag* to alert us when the transaction logs disk usage increased above a threshold or when a replication slot started lagging - meaning that Debezium might have died.
+* Heartbeats were not enabled. Heartbeats are needed to control the https://debezium.io/documentation/reference/1.0/connectors/postgresql.html#wal-disk-space[WAL disk space consumption] in the following cases:
+** There are many updates in a monitored database but only a minuscule amount relates to the monitored table(s) and/or schema(s). We handled this case by enabling heartbeats by setting `heartbeat.interval.ms`.
+** The PostgreSQL instance contains multiple databases where the monitored database is low-traffic in comparison to the others. Since the WAL is shared by all databases in an instance it keeps accumulating data which cannot be removed until Debezium reads it. But since the high traffic databases are not monitored Debezium is not able to communicate to the database that the WAL files can be removed to reclaim disk space. To solve this scenario we triggered "heartbeat" events by periodically updating a single row in a table created in the monitored database using the below query:
```

Comment:
You might add a link to https://issues.redhat.com/browse/DBZ-1815 to this section, which is about adding support for what you essentially built manually, into the connector. While I was initially shying away from it, adhering to a strong "no writes to source DB" policy, the issue is frequent enough, so that it warrants the built-in support for this (disabled by default of course).

Reply:
Thanks. Seems like the big guys are all using Debezium now, GitLab, Google Dataflow etc. Would be nice to see some insights from GitLab due to their scale.

Reply:
Oh, first time I'm hearing about GitLab using it. Do you have any references?

Reply:
Hmmmm, I think I mistook the `gitlab` suffix from a user on Gitter. 🤦‍♂ https://gitter.im/debezium/dev?at=5e4c80b08e726c7dc5b2ac82

Reply:
Ah, gotcha. That's people logging in using GitLab as identity provider.